apps/builder/app/builder/shared/assets/use-assets.tsx

```diff
@@ -269,65 +278,165 @@ const ToastMessage = ({
+      continue;
+    }
+
+    for (const fileData of filesData) {
+      const assetId = fileData.assetId;
+
+      if ($assets.get().has(assetId)) {
+        toast.warn(
+          <ToastMessage
+            title="Info"
+            assetId={assetId}
+            objectURL={fileData.objectURL}
+          >
+            Asset is already exists
+          </ToastMessage>
+        );
+
+        deleteUploadingFileData(assetId);
+        continue;
+      }
+
+      await uploadAsset({
+        authToken,
+        projectId,
+        fileOrUrl:
+          fileData.source === "file" ? fileData.file : new URL(fileData.url),
+        onCompleted: (data) => {
+          URL.revokeObjectURL(fileData.objectURL);
+          deleteUploadingFileData(assetId);
+          handleAfterSubmit(assetId, data, projectId);
+        },
+        onError: (error) => {
+          deleteUploadingFileData(assetId);
+          toast.error(
+            <ToastMessage
+              title="Error"
+              assetId={assetId}
+              objectURL={fileData.objectURL}
+            >
+              {error}
+            </ToastMessage>
+          );
+
+          safeDeleteAssets([assetId], projectId);
+        },
+      });
+    }
+  }
+};
+
+export async function uploadAssets(
+  type: AssetType,
+  files: File[]
+): Promise<Map<File, string>>;
+
+export async function uploadAssets(
+  type: AssetType,
+  urls: URL[]
+): Promise<Map<URL, string>>;
+
+/**
+ * returns a list of asset ids that are being uploaded
+ */
+// eslint-disable-next-line func-style
+export async function uploadAssets(
+  type: AssetType,
+  filesOrUrls: File[] | URL[]
+): Promise<Map<URL | File, string>> {
   const projectId = $project.get()?.id;
   const authToken = $authToken.get();
   if (projectId === undefined) {
-    return;
+    return new Map();
   }
 
-  const filesData = await getFilesData(type, files, urls);
+  const filesData = await getFilesData(type, filesOrUrls);
+
+  // Filter out duplicates inside filesData
+  const uniqFilesDataMap = new Map(
+    filesData.map((fileData) => [fileData.assetId, fileData])
+  );
 
-  addUploadingFilesData(filesData);
+  // Filter out duplicates existing in assets or uploading files
+  const existingIds = [
+    ...$assets.get().keys(),
+    ...$uploadingFilesDataStore.get().map((fileData) => fileData.assetId),
+  ];
 
-  for (const fileData of filesData) {
-    const assetId = fileData.assetId;
+  for (const existingAssetId of existingIds) {
+    if (uniqFilesDataMap.has(existingAssetId)) {
+      const fileData = uniqFilesDataMap.get(existingAssetId)!;
+      uniqFilesDataMap.delete(existingAssetId);
 
-    if ($assets.get().has(assetId)) {
       toast.warn(
         <ToastMessage
-          title="Info"
-          assetId={assetId}
+          title="Warning"
+          assetId={fileData.assetId}
           objectURL={fileData.objectURL}
         >
-          Asset is already exists
+          Asset is already being uploaded
```

Comment:
I see this message while the upload was already done, I am not sure what is meant here, that the image has been already uploaded? "being" is incorrect in this case, its already been uploaded